### ISSUE TITLE
Fatal error: concurrent map read and map write

### ISSUE THREAD

**Issue by tomasdeml** [mholt/caddy#1045]
#### 1. What version of Caddy are you running (`caddy -version`)?

v0.9.1
#### 2. What are you trying to do?

Refresh heavily a web site served by caddy reverse proxy, i.e. hold F5 key for a while in Chrome / Firefox.
#### 3. What is your entire Caddyfile?

``` text
https://:443 {
 tls self_signed
 proxy / 127.0.0.1:580 {
    transparent
    max_fails 0
  }
  bind 127.0.0.1
}
```
#### 4. How did you run Caddy (give the full command and describe the execution environment)?

`caddy.exe -conf Caddyfile`
#### 5. What did you expect to see?

I expected the web site to appear once the heavy refreshing is done.
#### 6. What did you see instead (give full error messages and/or log)?

The web site kept loading indefinitely (spinning wheel). Refresh or restart of browser did not help. Then Caddy failed with fatal error (see [attachment](https://github.com/mholt/caddy/files/425150/out2.zip)  for complete output):

``` text
Activating privacy features... done.
https:// (only accessible on this machine)
fatal error: concurrent map read and map write

goroutine 19217 [running]:
runtime.throw(0xa3a054, 0x21)
    /usr/local/go/src/runtime/panic.go:566 +0x9c fp=0xc0458ffc38 sp=0xc0458ffc18
runtime.mapaccess1_faststr(0x9ca040, 0xc04217d9b0, 0xa2aff8, 0xa, 0x94ce9aa0244a93f8)
    /usr/local/go/src/runtime/hashmap_fast.go:201 +0x4fa fp=0xc0458ffc98 sp=0xc0458ffc38
net/textproto.MIMEHeader.Get(0xc04217d9b0, 0xa2aff8, 0xa, 0xc04232eb00, 0x0)
    /usr/local/go/src/net/textproto/header.go:33 +0x7c fp=0xc0458ffcd0 sp=0xc0458ffc98
net/http.Header.Get(0xc04217d9b0, 0xa2aff8, 0xa, 0xa, 0xc04232eb88)
    /usr/local/go/src/net/http/header.go:39 +0x46 fp=0xc0458ffd08 sp=0xc0458ffcd0
net/http.(*Request).write(0xc042324d20, 0xcbbd20, 0xc04472c6c0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /usr/local/go/src/net/http/request.go:507 +0xabf fp=0xc0458ffe48 sp=0xc0458ffd08
net/http.(*persistConn).writeLoop(0xc042c71700)
    /usr/local/go/src/net/http/transport.go:1644 +0x1b3 fp=0xc0458fff98 sp=0xc0458ffe48
runtime.goexit()
    /usr/local/go/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc0458fffa0 sp=0xc0458fff98
created by net/http.(*Transport).dialConn
    /usr/local/go/src/net/http/transport.go:1058 +0x515
...
```
#### 7. How can someone who is starting from scratch reproduce this behavior as minimally as possible?

Unfortunately, I could not reproduce the issue by proxying to another Caddy instance or a public Internet site. I can only reproduce the issue with our application as upstream host running on localhost. It seems to be dependent on timing and number of requests from browser.

The issue does not occur when running caddy with HTTP2 disabled, when running without proxy or when running with proxy but without `max_fails` directive. I also could not reproduce the issue with `boom` benchmark tool (tried w/ and w/o `-h2` switch).

**Comment by mholt** [comment]
Can you post the full panic output? The stack trace is cut off...

**Comment by tomasdeml** [comment]
Sure, hope this [one](https://github.com/mholt/caddy/files/425147/out3.zip) is complete.

**Comment by mholt** [comment]
Got it, thanks. That's quite a busy server you've got there.

Weird, [this map access within the std lib](https://github.com/golang/go/blob/11e93aa24a61e0c4b25600bf2a681b8779371fe8/src/net/textproto/header.go#L33) is apparently concurrent with something else. So maybe there's something in our proxy.go or reverseproxy.go files that is accessing the same map at the same time? But how is that possible...?

Will look into this, would like some help as well.

**Comment by mholt** [comment]
Based on what you said:

> The issue does not occur when running caddy with HTTP2 disabled, when running without proxy or when running with proxy but without max_fails directive. 

It seems like the greatest common denominator is `max_fails 0` + HTTP/2. Remove either one of these and the race disappears, correct? What if max_fails is set to 1, I wonder? Meanwhile I'll try reproducing this with wrk...

**Comment by tomasdeml** [comment]
> Remove either one of these and the race disappears, correct?
Yep.

I will be out of Office for a while so I cannot try it right now with `max_fails 1` :(

**Comment by mholt** [comment]
Dang. Well I would really love to be able to reproduce this. In the meantime, if anyone can examine the proxy.go and reverseproxy.go files and see if there is shared map state somewhere in there, that would be greatly appreciated.

**Comment by zacheryph** [comment]
https://gist.github.com/anonymous/146fe3bcaa53505978954a412324f911

`fatal error: concurrent map writes`

I received this on `Caddy 0.9.0`.

**Comment by mholt** [comment]
@zacheryph So yours is having concurrent map writes [here](https://github.com/golang/go/blob/da6b9ec7bf1722fa00196e1eadc10a29156b6b28/src/net/textproto/header.go#L22) and [here](https://github.com/mholt/caddy/blob/f28af637327a4f12ae745284c519cfdeca5502ef/caddyhttp/proxy/proxy.go#L132). We've taken out that code in 0.9.1 - can you upgrade and post a stack trace if you still see a fault?

I'm not sure still the source of your panic, @tomasdeml - it could possibly be that we're passing in host.UpstreamHeaders [here](https://github.com/mholt/caddy/blob/fdc62d015fd5c20e03143e2eea17528892270169/caddyhttp/proxy/proxy.go#L129) for reading, however this line doesn't appear in your stack trace, and your panic reports a concurrent _read and write_. Our function doesn't write host.UpstreamHeaders there. The weird thing about that stack trace is the whole thing is in the std lib. The read is [probably this line from the stack trace](https://github.com/golang/go/blob/759c2c48ea14d3b9c9ee594862472761b706c98f/src/net/textproto/header.go#L33) but I can't figure out where the write would be coming from...

**Comment by zacheryph** [comment]
Yeah, I will rebuild my container and if it happens again get a log out.

**Comment by tw4452852** [comment]
It seems we reuse the `request` when retrying:

```
goroutine 41104 [running]:
runtime.throw(0xa3a054, 0x21)
        /usr/local/go/src/runtime/panic.go:566 +0x9c fp=0xc049709c38 sp=0xc049709c18
runtime.mapaccess1_faststr(0x9ca040, 0xc04830a930, 0xa2aff8, 0xa, 0x60c8ef66792824da)
        /usr/local/go/src/runtime/hashmap_fast.go:201 +0x4fa fp=0xc049709c98 sp=0xc049709c38
net/textproto.MIMEHeader.Get(0xc04830a930, 0xa2aff8, 0xa, 0xc042338000, 0x0)
        /usr/local/go/src/net/textproto/header.go:33 +0x7c fp=0xc049709cd0 sp=0xc049709c98
net/http.Header.Get(0xc04830a930, 0xa2aff8, 0xa, 0xa, 0xc0423380b8)
        /usr/local/go/src/net/http/header.go:39 +0x46 fp=0xc049709d08 sp=0xc049709cd0
net/http.(*Request).write(0xc0420b00f0, 0xcbbd20, 0xc0483949c0, 0xc049753f00, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/http/request.go:507 +0xabf fp=0xc049709e48 sp=0xc049709d08
net/http.(*persistConn).writeLoop(0xc0446f2f00)
        /usr/local/go/src/net/http/transport.go:1644 +0x1b3 fp=0xc049709f98 sp=0xc049709e48
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc049709fa0 sp=0xc049709f98
created by net/http.(*Transport).dialConn
        /usr/local/go/src/net/http/transport.go:1058 +0x515
```

and this:

```
goroutine 334 [runnable]:
github.com/mholt/caddy/caddyhttp/httpserver.(*replacer).Replace(0xc04234b050, 0xa29e2d, 0x8, 0x1, 0x0)
        /Users/matt/Dev/src/github.com/mholt/caddy/caddyhttp/httpserver/replacer.go:224 +0x4a3
github.com/mholt/caddy/caddyhttp/proxy.mutateHeadersByRules(0xc0423ce5d0, 0xc042178a80, 0xcc0ca0, 0xc04234b050)
        /Users/matt/Dev/src/github.com/mholt/caddy/caddyhttp/proxy/proxy.go:241 +0x34d
github.com/mholt/caddy/caddyhttp/proxy.Proxy.ServeHTTP(0xcbeb20, 0xc042178bd0, 0xc0420e8c10, 0x1, 0x1, 0xcc30e0, 0xc0420e60d0, 0xc042243770, 0x6, 0x6, ...)
        /Users/matt/Dev/src/github.com/mholt/caddy/caddyhttp/proxy/proxy.go:129 +0x2af
github.com/mholt/caddy/caddyhttp/proxy.(*Proxy).ServeHTTP(0xc042178c00, 0xcc30e0, 0xc0420e60d0, 0xc042243770, 0xc042243701, 0x0, 0x1c)
        <autogenerated>:5 +0x96
github.com/mholt/caddy/caddyhttp/httpserver.(*Server).serveHTTP(0xc04223bd50, 0xcc30e0, 0xc0420e60d0, 0xc042243770, 0x5, 0x0, 0x0)
        /Users/matt/Dev/src/github.com/mholt/caddy/caddyhttp/httpserver/server.go:267 +0x234
github.com/mholt/caddy/caddyhttp/httpserver.(*Server).ServeHTTP(0xc04223bd50, 0xcc30e0, 0xc0420e60d0, 0xc042243770)
        /Users/matt/Dev/src/github.com/mholt/caddy/caddyhttp/httpserver/server.go:214 +0xec
net/http.serverHandler.ServeHTTP(0xc04218a380, 0xcc30e0, 0xc0420e60d0, 0xc042243770)
        /usr/local/go/src/net/http/server.go:2202 +0x84
net/http.initNPNRequest.ServeHTTP(0xc042292000, 0xc04218a380, 0xcc30e0, 0xc0420e60d0, 0xc042243770)
        /usr/local/go/src/net/http/server.go:2670 +0x9a
net/http.(*initNPNRequest).ServeHTTP(0xc0421805a0, 0xcc30e0, 0xc0420e60d0, 0xc042243770)
        <autogenerated>:288 +0x80
net/http.(Handler).ServeHTTP-fm(0xcc30e0, 0xc0420e60d0, 0xc042243770)
        /usr/local/go/src/net/http/h2_bundle.go:4084 +0x54
net/http.(*http2serverConn).runHandler(0xc0420118c0, 0xc0420e60d0, 0xc042243770, 0xc04223dee0)
        /usr/local/go/src/net/http/h2_bundle.go:4314 +0x90
created by net/http.(*http2serverConn).processHeaders
        /usr/local/go/src/net/http/h2_bundle.go:4092 +0x6e9
```

**Comment by mholt** [comment]
Thanks Tw, that could be the case although I'm still trying to understand it fully.

The stack trace shows:

```
github.com/mholt/caddy/caddyhttp/proxy.Proxy.ServeHTTP(0x7f03a7f92450, 0xc820180f30, 0xc820107340, 0x1, 0x1, 0x7f03a813e680, 0xc8203d8000, 0xc8201b6620, 0x7c142e, 0x0, ...)
 	/tmp/custombuild_4845_729452719/src/github.com/mholt/caddy/caddyhttp/proxy/proxy.go:132 +0x703 fp=0xc82055b600 sp=0xc82055b3d8
```

but [proxy.go line 132 is a closing curly brace](https://github.com/mholt/caddy/blob/c5aa5843d92a27eaf521e28684111030135d9cdc/caddyhttp/proxy/proxy.go#L132) on 0.9.1. How did that make it into the stack trace? O.o

**Comment by tw4452852** [comment]
@mholt I mean [this request](https://github.com/mholt/caddy/blob/master/caddyhttp/proxy/proxy.go#L90)

**Comment by mholt** [comment]
@tw4452852 Okay, so outreq is being reused, this is true. That must mean that the standard lib is accessing its header map concurrently somewhere...? Maybe we should make a new outreq in each iteration of the loop instead?

Tricky since we can't easily reproduce this :( I have yet to be able to trigger it.

**Comment by tw4452852** [comment]
@mholt I think it's because we reuse the `outreq` when retrying, but the standard lib may still use it after `RoundTrip`. So we could make a new `outreq` before each iteration of retrying loop for now.

BTW, just quote from the doc of `RoundTrip`:

> 	// RoundTrip must always close the body, including on errors,
	// but depending on the implementation may do so in a separate
	// goroutine even after RoundTrip returns. This means that
	// callers wanting to reuse the body for subsequent requests
	// must arrange to wait for the Close call before doing so.

So I think possibility of reuse may exists...

Also I think this issue only reproduced on http2 path and also forward the request to proxy.

**Comment by mholt** [comment]
If we move the creation of outreq to inside the loop, we should comment describing why we're doing that (with a link to this issue).

@tomasdeml Are you in a position to try a change once we have one ready?

**Comment by tomasdeml** [comment]
@mholt Yep, I'm at your disposal now. I will also try to reproduce the issue with `max_fails 1`.

**Comment by tomasdeml** [comment]
With `max_fails 1` I start getting `502 backend unreachable` from Caddy after a few seconds, but no panic error. I also tried `max_fails 0` with `fail_timeout 1s` and the panic does not occur.

**Comment by mholt** [comment]
@tomasdeml Do you want to dive into this more and try that code change locally and see if it works? Otherwise I'll leave this as a Hacktoberfest item, I just don't have the time right now to do it.

**Comment by tomasdeml** [comment]
@mholt We have increased `max_fails` to 5 (to prevent spinning) and disabled HTTP2 (as we do not require it presently). So the issue can be dealt with later :-)

**Comment by mholt** [comment]
Lots of things have changed in the proxy middleware in the last 12 months. I haven't heard any similar reports of this happening (nor have I seen it myself) so it's possible that the bug is fixed? Also, I'm rewriting the proxy middleware and am trying to be careful about this kind of thing. We'll see how it goes. But I think I'll close this, for lack of activity on it.

**Comment by mholt** [comment]
@faradaytrs Please open a new issue, and fill out the issue template. Thanks!

**Comment by christophcemper** [comment]
@mholt this looks like a very ancient issue, but we saw this surface first time ever on production Oct 28 and also last night again Nov 9
on a very busy server.

what do you suggest to monitor/enable in logs/etc to try to capture sufficient info for investigating? we've no idea how to reproduce infortunately.

![Image](https://github.com/user-attachments/assets/b4e18f56-14a0-454f-8685-82f1f4ec3a77)

**Comment by mholt** [comment]
@christophcemper Please open a new issue, this one is eight years old, and is a totally separate issue.
